modules/mqtt/destination/mqtt-destination.c

```diff
@@ -0,0 +1,280 @@
+void
+mqtt_destination_dd_set_topic(LogDriver *d, const gchar *topic)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  g_string_assign(self->topic, topic);
+}
+
+
+void
+mqtt_destination_dd_set_keepalive (LogDriver *d, const gint keepalive)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  self->keepalive = keepalive;
+}
+
+void
+mqtt_destination_dd_set_qos (LogDriver *d, const gint qos)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  self->qos = qos;
+}
+
+void
+mqtt_destination_dd_set_protocol_type(LogDriver *d, const gchar *protocol_type)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  g_string_assign(self->protocol_type, protocol_type);
+}
+
+void
+mqtt_destination_dd_set_message_ref(LogDriver *d, LogTemplate *message)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  log_template_unref(self->message);
+  self->message = message;
+}
+
+/*
+ * Utilities
+ */
+static const gchar *
+_format_stats_instance(LogThreadedDestDriver *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar stats_instance[1024];
+
+  if (((LogPipe *)d)->persist_name)
+    g_snprintf(stats_instance, sizeof(stats_instance), "%s", ((LogPipe *)d)->persist_name);
+  else
+    g_snprintf(stats_instance, sizeof(stats_instance),
+              "mqtt-destination,%s,%d,%s", self->host->str, self->port, self->topic->str);
+  return stats_instance;
+}
+
+static const gchar *
+_format_persist_name(const LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar persist_name[1024];
+
+  if (d->persist_name)
+    g_snprintf(persist_name, sizeof(persist_name), "mqtt-destination.%s", d->persist_name);
+  else
+    g_snprintf(persist_name, sizeof(persist_name), "mqtt-destination.(%s,%d,%s)", self->host->str, self->port, self->topic->str);
+
+  return persist_name;
+}
+
+static gboolean
+_check_is_valid_protocol(gchar *protocol_type)
+{
+  gchar* valid_type[] = {"tcp", "ssl", "ws", "wss"};
+  gint i;
+
+  for (i = 0; i < G_N_ELEMENTS(valid_type); ++i)
+      if (strcmp(protocol_type, valid_type[i]) == 0)
+        return TRUE;
+
+  return FALSE;
+}
+
+static void
+_set_default_value(LogPipe *s)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)s;
+  GlobalConfig *cfg = log_pipe_get_config(s);
+
+  if (self->host->len == 0)
+    g_string_assign(self->host, DEFAULT_HOST);
+
+  if (self->port == -1)
+    self->port = DEFAULT_PORT;
+
+  if (self->keepalive == -1)
+    self->keepalive = DEFAULT_KEEPALIVE;
+
+  if (self->qos == -1)
+    self->qos = DEFAULT_QOS;
+
+  if (self->protocol_type->len == 0)
+    g_string_assign(self->protocol_type, DEFAULT_PROTOCOL_TYPE);
+  
+  if (self->message == NULL)
+  {
+    self->message = log_template_new(cfg, NULL);
+    log_template_compile(self->message, DEFAULT_MESSAGE_TEMPLATE, NULL);
+  }
+
+  log_template_options_init(&self->template_options, cfg);
+}
+
+static gboolean
+_dd_init(LogPipe *d)
```

Comment:
since we are not going to put the source in the same file and this is even in a separate "destination" subdirectory, I would completely drop the "_dd" prefix.

Reply:
Solved by this comment: MQTT/destination: remove _dd and _dw prefix
(commit hash: 3bdd0a9f7bbf4e148a7a7cc16a7c782a9556cae7 )